### ISSUE TITLE
Correct behaviour when switching `Player` in `MediaSession`

### ISSUE THREAD

**Issue by NielsMasdorp** [androidx/media#39]
As per #25 i've implemented a custom `Player` which caches the `MediaItem`(s) internally and switches from local playback using an `ExoPlayer` instance to a `CastPlayer` instance when a cast session is available and vice versa. This composite `Player` passed to the `MediaSession` and no subsequent calls to `MediaSession#setPlayer` are made.

```
player = CompositePlayer(localPlayer = localPlayer,castPlayer = castPlayer)
mediaSession = MediaSession.Builder(this, player)
```

The way I am currently switching players internally is: stop the playback of the current player using `stop()` and clear the playlist in the current player using `clearMediaItems()`. Afterwards set the items to the new player using `#setMediaItem`, `prepare()` and `play()`. 

This all works fine, except when switching from local playback to cast playback a new cast notification is created, yet the "old" media notification is also still present, so theres 2 notifications at the same time, also the "old" notification is not updated anymore (presumably because it is connected to the old local playback which is not used anymore?).

My question: what is supposed to be the correct UX? I feel like there should only be 1 notification: a cast notification when casting and 1 notification when using local playback.

**Comment by marcbaechinger** [comment]
What would be the preferred option for you as a developer?

I would want to have a single place where I can configure how the notification looks like and whats written into the media session. Being an Android app I would think I do that with the media session that I need to maintain with my Android app. If there is a cast device involved temporarily I would want that I don't have to do the same customisation and session maintaining again. 

I wouldn't care from where this session and notification is coming, but because I develop an Android app I need to maintain a session and a notification as long as my Android app is alive. The Cast session and notification is not helpful here because it can't manage my session/notification when I'm not casting.

I would also expect that someone who listens to my media session while my local playback is running doesn't have to know that my app may create a second session while I'm casting. So I want to have consistently the same session for my app.

So for the reasons above I as a developer would want that Media3 is maintaining a single session and a single notification for the lifetime of my app. That simplest for myself as a developer and it is less confusing by an external app that wants to interact with the session of my app.

I would be interest to hear you opinion regarding this.

I would also be interested whether it helps if you change you `CastOptionsProvider` and add there:

```
.setCastMediaOptions(
    CastMediaOptions.Builder()
         .setMediaSessionEnabled(false)
         .setNotificationOptions(null)
         .build()
)
```

**Comment by NielsMasdorp** [comment]
Yes, I would agree that from a developer perspective it is logical to maintain a single `MediaSession` which in turn manages a single notification which can switch between a "cast notification" and a "local notification". However I am not sure what a "cast notification" should look like.

However right now the following is happening:

**Local playback (default state without initialising a cast session):**

![image](https://user-images.githubusercontent.com/5968599/154250837-82a6b8eb-3fc4-472f-9409-29d9a57b0638.png)

**After initialising cast session and thus resetting the `ExoPlayer` instance in the Composite `Player` and starting the `CastPlayer` another notification is shown and the "old" notification changes state (note the badge which specifies the casting device):**

![image](https://user-images.githubusercontent.com/5968599/154250996-a46d51f4-6641-40b2-b50d-650dd2c246a4.png)
_kind of confusing the cast notification says "other device"_

![image](https://user-images.githubusercontent.com/5968599/154251020-daba130a-bc64-4fc2-ab80-3dcca132ecf5.png)

Imo it is kind of confusing to have 2 notifications with playback controls and one notification with a close (which stops the casting session). 

When I disable the mediasession and notification options in the `OptionsProvider` the result is as expected:

There is always 1 notification showing:

![image](https://user-images.githubusercontent.com/5968599/154251809-d480da4f-d87c-4161-82f2-937d1a876ceb.png)

However that notification does not show the google cast logo next to the artwork and there is no close button to stop the casting session. Which, from an user perspective, is unexpected.

What I would expect is: always 1 notification and with local playback its the normal notification:

![image](https://user-images.githubusercontent.com/5968599/154252102-9a4bfc86-ad7f-40f1-9b5b-f4104825f9f5.png)

And when I switch to cast its only this notification:

![image](https://user-images.githubusercontent.com/5968599/154252183-5dfc0c27-2683-4acf-a61c-908c19800c1f.png)

However I am not sure this is possible when using a composite player in the `MediaSession` which switches players internally. Maybe it's only possible when we switch players in the `MediaSession` using `MediaSession#setPlayer` on the fly?

As extra clarification: I am not creating/managing any notifications myself, these notifications are all created by media3 internally.

**Comment by NielsMasdorp** [comment]
@marcbaechinger Thinking about it I could probably add a custom action to the notification when a cast session is active and sort of have the same effect was going for in the first place. And reach our preferable outcome: one single notification and one [MediaSession] per app.

**Comment by AntonMishanin** [comment]
Hello!

I have the same issue with duplicate notifications. I fixed it by:
```
.setCastMediaOptions(
    CastMediaOptions.Builder()
         .setMediaSessionEnabled(false)
         .setNotificationOptions(null)
         .build()
)
```
But after this I can't change volume on TV via hardware buttons - remote volume is always at zero (as if the volume is turned off) and does not change
Without `.setNotificationOptions(null)` volume changes works fine
![2025-01-14 17 08 09](https://github.com/user-attachments/assets/e157a397-a46f-4216-bf51-af85feeec37c)


What do you think about it? How manage the media session and the notifications ourselves and change remote volume?
@marcbaechinger

**Comment by AquilesCanta** [comment]
Hello folks, I know it's been a while since this was reported but I wanted to flag:

> i've implemented a custom Player which caches the MediaItem(s) internally and switches from local playback using an ExoPlayer instance to a CastPlayer instance when a cast session is available and vice versa

We've made CastPlayer do this in version 1.9. That is: A player that handles both local and Cast playback. Don't worry: We've kept the old CastPlayer constructors' behavior exactly as it was. You get this Exo+Cast playback composition in a single player implementation if you create CastPlayer using the builder. If you still want a player that only supports Cast, you can use RemoteCastPlayer, which is what CastPlayer used to be in 1.8 and older. Further details in https://developer.android.com/media/media3/cast/create-castplayer. I'm sorry you had to wait so long for this.

> This all works fine, except when switching from local playback to cast playback a new cast notification is created, yet the "old" media notification is also still present

The right fix for this is mentioned in https://github.com/androidx/media/issues/39#issuecomment-2590103836. You need to set the cast option 

```
.setCastMediaOptions(
    CastMediaOptions.Builder()
         .setMediaSessionEnabled(false) <---- this line
```
